scripts/cobot280pi/client.py

```diff
@@ -0,0 +1,147 @@
+import paho.mqtt.client as paho
+import time
+import json
+import threading
+import matplotlib.pyplot as plt
+from queue import Queue
+import numpy as np
+from PIL import Image
+
+response_queues = {}
+
+def on_connect(client, userdata, flags, rc, properties=None):
+	print("Connection recieved")
+
+# with this callback you can see if your publish was successful
+def on_publish(client, userdata, mid, properties=None):
+	print("mid: " + str(mid))
+
+def on_message(client, userdata, msg):
+	if msg.topic not in response_queues:
+		response_queues[msg.topic] = Queue()
+	response_queues[msg.topic].put(msg)
+
+class CobotMQTTClient:
+    
+    def __init__(
+		self,
+		hive_mq_username: str,
+		hive_mq_password: str,
+		hive_mq_cloud: str,
+		port: int = 8883
+	):
+        self.client = paho.Client(client_id="", userdata=None, protocol=paho.MQTTv5)
+        self.client.on_connect = on_connect
+        # client.on_subscribe = on_subscribe
+        self.client.on_message = on_message
+        self.client.on_publish = on_publish
+
+        self.client.tls_set(tls_version=paho.ssl.PROTOCOL_TLS)
+        self.client.username_pw_set(hive_mq_username, hive_mq_password)
+        self.client.connect(hive_mq_cloud, port)
+
+        self.client.subscribe("response/query/angles", qos=2)
+        self.client.subscribe("response/query/coords", qos=2)
+        self.client.subscribe("response/query/camera", qos=2)
+        self.client.subscribe("response/control/angles", qos=2)
+        self.client.subscribe("response/control/coords", qos=2)
+        self.client.subscribe("response/control/gripper", qos=2)
+        
+        self.client_loop_thread = threading.Thread(target=self.client.loop_forever)
```

Comment:
I'm curious, why threading here?

Reply:
From what I understood of paho-mqtt, this function needs to be run for messages to be received.

I set it up so that the client could be interactive (like in the REPL or a Jupyter kernel, so I couldn't run the blocking function without threading.

Not sure if there is a better way for this to be done.

Reply:
Thanks for clarifying! I think I see your point, although there may be some issues with Streamlit and hugging face spaces down the line. Two alternatives are asynchronous and non-threaded synchronous. However I think you need to use a different library for asynchronous. Non-threaded synchronous might be the best starting point. Some of my musings: https://chatgpt.com/share/66ef1b93-80d0-8006-8996-e1d84a9bbea3

Reply:
@sgbaird removed threading from client in the most recent push, let me know how it looks now